src/jdk.compiler/share/classes/com/sun/tools/javac/code/LintMapper.java

```diff
@@ -0,0 +1,329 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.tools.javac.code;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import javax.tools.DiagnosticListener;
+import javax.tools.JavaFileObject;
+
+import com.sun.tools.javac.tree.EndPosTable;
+import com.sun.tools.javac.tree.JCTree;
+import com.sun.tools.javac.tree.JCTree.*;
+import com.sun.tools.javac.tree.TreeInfo;
+import com.sun.tools.javac.tree.TreeScanner;
+import com.sun.tools.javac.util.Assert;
+import com.sun.tools.javac.util.Context;
+import com.sun.tools.javac.util.JCDiagnostic.DiagnosticPosition;
+
+/**
+ * Maps source code positions to the applicable {@link Lint} instance.
+ *
+ * <p>
+ * Because {@code @SuppressWarnings} is a Java symbol, in general this mapping can't be
+ * calculated until after attribution. As each top-level declaration (class, package, or module)
+ * is attributed, this singleton is notified and the {@link Lint}s that apply to every source
+ * position within that top-level declaration are calculated.
+ *
+ * <p>
+ * The method {@link #lintAt} returns the {@link Lint} instance applicable to source position;
+ * if it can't be determined yet, an empty {@link Optional} is returned.
+ *
+ * <p><b>This is NOT part of any supported API.
+ * If you write code that depends on this, you do so at your own risk.
+ * This code and its internal interfaces are subject to change or
+ * deletion without notice.</b>
+ */
+public class LintMapper {
+
+    // The key for the context singleton
+    private static final Context.Key<LintMapper> CONTEXT_KEY = new Context.Key<>();
+
+    // Per-source file information. Note: during the parsing of a file, an entry exists but the FileInfo value is null
+    private final Map<JavaFileObject, FileInfo> fileInfoMap = new HashMap<>();
```

Comment:
would it makes sense to make this map a WeakHashMap. I'm thinking about cases where we are not using a ReusableContext and thus this map wont be cleared.

Reply:
yes you are right we have other DSs with the same `issue` if any. I guess discuss this and address it as a separate issue